vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,797 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace rec Microsoft.VisualStudio.FSharp.Editor
+
+open System
+open System.Windows.Controls
+open System.Windows.Media
+open Microsoft.VisualStudio.Text
+open Microsoft.VisualStudio.Text.Editor
+open Microsoft.VisualStudio.Text.Formatting
+open System.ComponentModel.Composition
+open Microsoft.VisualStudio.Utilities
+open Microsoft.CodeAnalysis
+open System.Threading
+open Microsoft.FSharp.Compiler.SourceCodeServices
+open Microsoft.VisualStudio.Shell
+open Microsoft.VisualStudio
+open Microsoft.VisualStudio.LanguageServices
+open System.Windows
+open System.Collections.Generic
+open Microsoft.FSharp.Compiler.Range
+open Microsoft.FSharp.Compiler
+open Microsoft.FSharp.Compiler.Ast
+open Microsoft.CodeAnalysis.Editor.Shared.Extensions
+open Microsoft.CodeAnalysis.Editor.Shared.Utilities
+open Microsoft.CodeAnalysis.Classification
+open Internal.Utilities.StructuredFormat
+open Microsoft.VisualStudio.Text.Tagging
+open System.Collections.Concurrent
+open System.Collections
+open System.Windows.Media.Animation
+open System.Globalization
+
+open Microsoft.VisualStudio.FSharp.Editor.Logging
+open Microsoft.CodeAnalysis.Text
+
+type CodeLensGeneralTag(width, topSpace, baseline, textHeight, bottomSpace, affinity, tag:obj, providerTag:obj) =
+    inherit SpaceNegotiatingAdornmentTag(width, topSpace, baseline, textHeight, bottomSpace, affinity, tag, providerTag)
+
+/// Class which provides support for general code lens
+/// Use the methods <code>AddCodeLens</code< and <code> RemoveCodeLens</code>
+type CodeLensGeneralTagger (buffer:ITextBuffer) as self =
+    // Custom simple tagger logic
+
+    /// Saves the ui context to switch context for ui related work.
+    let uiContext = SynchronizationContext.Current
+
+    /// The tags changed event to notify if the data for the tags has changed.
+    let tagsChangedEvent = new Event<EventHandler<SnapshotSpanEventArgs>,SnapshotSpanEventArgs>()
+
+    // Tracks the created ui elements per TrackingSpan
+    let mutable uiElements = Dictionary<_,_>()
```

Comment:
Dictionary objects are always mutable - you would only use ``mutable`` here if you were doing ``uiElements <- ...`` at some point.